### PR TITLE
kodi: 19.4 -> 20.0

### PR DESCRIPTION
**Pull request by aanderse**
###### Description of changes

New major version of `kodi` ~~is now in `beta`~~ has been released. We can use this PR to coordinate testing.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasj19** [comment]
Thanks for packaging ahead of stable. I'm building this right now and will keep an eye on how it works

**Comment by aanderse** [comment]
> I'm building this right now and will keep an eye on how it works

I appreciate that but I'll put the caveat that it won't work well *yet*. I have a bunch of commits to add to this PR which I haven't got around to pushing yet. I'll try to do that over the weekend and ping you when done.

Why haven't I pushed yet? Waiting for [this](https://github.com/NixOS/nixpkgs/pull/201836) to be merged... so I need to test it first. Again... hopefully this weekend :crossed_fingers:

**Comment by dasj19** [comment]
@aanderse I'm offering again to test version 20, I see that upstream is no longer in beta.
Feel free to ping me when this is ready for testing. Maybe also specify what should be in focus.

**Comment by aanderse** [comment]
@dasj19 thank you for the kind offer. I haven't tested anything yet and will leave it in your capable hands 🙇‍♂️

File: pkgs/applications/video/kodi/addons/invidious/default.nix [hunk 1]
```diff
@@ -6,7 +6,7 @@ buildKodiAddon rec {
   version = "0.1.0+matrix.1";
 
   src = fetchzip {
-    url = "https://mirrors.kodi.tv/addons/matrix/${namespace}/${namespace}-${version}.zip";
+    url = "https://mirrors.kodi.tv/addons/nexus/${namespace}/${namespace}-${version}.zip";
```
Comment (dasj19) [inline pkgs/applications/video/kodi/addons/invidious/default.nix:9]: While not a packaging issue, this addon has several problems:
1. the default invidious URL is no longer active (has been fixed upstream: https://github.com/TheAssassin/kodi-invidious-plugin/commit/5474b454d20f16e924a936a49e1c7735f5d9bacb )
2. the search does not work, so one can only see Trending and Popular videos

All in all, I don't know if this module is worth keeping

**Comment by dasj19** [comment]
I am testing with the following config (to include specifically everything that was changed in this PR):

```
    (pkgs.kodi.passthru.withPackages (kodiPkgs: with kodiPkgs; [
      # test
      arrow arteplussept certifi chardet dateutil defusedxml future idna
      inputstream-adaptive inputstream-ffmpegdirect inputstream-rtmp
      inputstreamhelper invidious joystick keymap six
      libretro-genplus libretro-mgba libretro-snes9x
      libretro myconnpy pvr-hdhomerun pvr-hts pvr-iptvsimple requests-cache
      requests routing signals simplejson six steam-controller trakt-module
      trakt typing_extensions urllib3 vfs-libarchive vfs-sftp
      visualization-waveform websocket youtube
    ]))
```
I will post the issues I encounter

File: pkgs/applications/video/kodi/addons/inputstream-adaptive/default.nix [hunk 2]
```diff
@@ -1,16 +1,29 @@
 { stdenv, lib, rel, addonDir, buildKodiBinaryAddon, fetchFromGitHub, expat, glib, nspr, nss, gtest }:
+let
+  bento4 = fetchFromGitHub {
```
Comment (SuperSandro2000) [inline pkgs/applications/video/kodi/addons/inputstream-adaptive/default.nix:3]: Why not use our https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/video/bento4/default.nix ?

Comment (aanderse) [reply pkgs/applications/video/kodi/addons/inputstream-adaptive/default.nix:3]: - our `bento4` doesn't resemble what upstream distributes so this isn't currently an option, though i did some work on that and could fix it ...if i find motivation to do so
- even if i were to fix the above point `kodi` is nefarious for vendoring their own modified dependencies for years and years on end - i took a look and what `kodi` has for `bento4` isn't even close to the upstream release :disappointed:

Comment (peterhoeg) [reply pkgs/applications/video/kodi/addons/inputstream-adaptive/default.nix:3]: The proper fix (and I'm playing fast and loose with "proper" here) is to override `src` on the nixpkgs `bento4` package and build that normally - then have the plugin build with an external bento4 and pass it the custom one. This allows us to cache the bento4 package and benefit for what custom logic we currently have in nixpkgs's bento4.

Comment (aanderse) [reply pkgs/applications/video/kodi/addons/inputstream-adaptive/default.nix:3]: > * our `bento4` doesn't resemble what upstream distributes so this isn't currently an option, though i did some work on that and could fix it ...if i find motivation to do so

I guess I need to emphasize this more. Our `bento4` package looks **nothing** like what upstream (or `kodi`l provides. The entire directory structure is different and there are almost no overlapping files. The way we build `bento4` is almost completely incompatible (wrong?). There is nothing of value in our `bento4` package for `kodi`.

If course of I go ahead and "fix" our `bento4` then what you said is entirely accurate, which is why you said it 👍

I'll try to find some motivation to finish my `bento4` work (very close) and then make a PR... at some point.

On that note I would also like to clean up how we build all these `kodi` vendored packages and do the same for the rest of the dependencies we don't do this for yet.

Let's leave this comment unresolved as an indicator I'll come back to this at some point in the future after this PR is merged.

**Comment by aanderse** [comment]
@dasj19 any news on testing?

**Comment by dasj19** [comment]
Hello, sorry for the long delay, I have used the new kodi for a couple of hours total time... I did not encounter any packaging issue. One of my friends detected some issues with the subtitle rendering of fonts but that is because the upstream changed the subtitle rendering engine. So I guess we are good to go.

**Comment by aanderse** [comment]
Fantastic! Thank you so much. I'll resolve conflicts and merge hopefully tonight.

**Comment by LeSuisse** [comment]
Flagging as security related, this upgrade contains the fix for CVE-2023-23082.

**Comment by aanderse** [comment]
Thanks for the ping @LeSuisse! Sorry it took so long... rebased and hopefully this should be good to merge once the checks pass.

**Pull request merged**